File: optimizers/Sophia.py
from .Optimizer import Optimizer, np


class Sophia(Optimizer):
    """
    Оптимизатор Sophia (Second-order Clipped Stochastic Optimization)
    из статьи "Sophia: A Scalable Stochastic Second-order Optimizer
    for Language Model Pre-training" (arXiv:2305.14342)

    Вариант SophiaH, но вместо стохастической оценки Хатчинсона
    используется диагональ точного численного гессиана function.hesse,
    доступного в этой песочнице
    """

    m: np.ndarray
    h: np.ndarray

    def __init__(
        self,
        initial_x: np.ndarray,
        function,
        lr: float = 0.5,
        beta1: float = 0.96,
        beta2: float = 0.99,
        rho: float = 1.0,
        eps: float = 1e-12,
    ) -> None:
        params = dict(lr=lr, beta1=beta1, beta2=beta2, rho=rho, eps=eps)
        super().__init__(initial_x, function, params)

    def next_point(self) -> tuple[np.ndarray, float]:
        gradient = self.function.grad(self.x)
        # градиент точный и стохастики нет, поэтому оценку гессиана
        # обновляем каждый шаг (k = 1), а не раз в k шагов, как в статье
        hesse_diag = np.diag(self.function.hesse(self.x))
        self.m = self.params["beta1"] * self.m + (1 - self.params["beta1"]) * gradient
        # отрицательную кривизну обрезаем нулём, чтобы не шагать в сторону седла или максимума
        self.h = self.params["beta2"] * self.h + (1 - self.params["beta2"]) * np.maximum(hesse_diag, 0)
        # покомпонентный клиппинг порогом rho: шаг по каждой оси не длиннее lr * rho,
        # что защищает от взрыва, пока оценка кривизны h ещё мала
        update = np.clip(self.m / np.maximum(self.h, self.params["eps"]), -self.params["rho"], self.params["rho"])
        next_x = self.x - self.params["lr"] * update
        return self.move_next(next_x)

    def reset(self) -> None:
        super().reset()
        self.m = np.zeros([2])
        self.h = np.zeros([2])

File: src/shared/lib/optimization-engine/optimizers/sophia.ts
import type { OptimizerDescriptor } from "./types";
import { zero2, type Vec2 } from "@shared/lib/optimization-engine/linalg";
import { gradient, hessian } from "@shared/lib/optimization-engine/functions";

export const sophiaOptimizer: OptimizerDescriptor = {
  name: "Sophia",
  params: {
    lr: { default: 0.5, description: "скорость обучения — масштаб шага" },
    beta1: { default: 0.96, description: "коэффициент EMA первого момента (сглаживание направления градиента)" },
    beta2: {
      default: 0.99,
      description: "коэффициент EMA диагонали гессиана (оценка кривизны; отрицательная кривизна обрезается нулём)",
    },
    rho: {
      default: 1.0,
      description: "порог покомпонентного клиппинга шага — максимальная длина шага по оси равна lr * rho",
    },
    eps: { default: 1e-12, description: "нижний порог оценки кривизны h в знаменателе — численная защита" },
  },
  createInstance(fn, initialX, params) {
    let x: Vec2 = initialX;
    let m: Vec2 = zero2();
    let h: Vec2 = zero2();
    return {
      get x() { return x; },
      params,
      next() {
        const grad = gradient(fn, x[0], x[1]);
        const hesse = hessian(fn, x[0], x[1]);
        const hesseDiag: Vec2 = [hesse[0][0], hesse[1][1]];
        m = [
          params.beta1 * m[0] + (1 - params.beta1) * grad[0],
          params.beta1 * m[1] + (1 - params.beta1) * grad[1],
        ];
        h = [
          params.beta2 * h[0] + (1 - params.beta2) * Math.max(hesseDiag[0], 0),
          params.beta2 * h[1] + (1 - params.beta2) * Math.max(hesseDiag[1], 0),
        ];
        const update: Vec2 = [
          Math.min(Math.max(m[0] / Math.max(h[0], params.eps), -params.rho), params.rho),
          Math.min(Math.max(m[1] / Math.max(h[1], params.eps), -params.rho), params.rho),
        ];
        x = [x[0] - params.lr * update[0], x[1] - params.lr * update[1]];
        return { x, value: fn(x[0], x[1]), internals: {
          "m.x": m[0], "m.y": m[1], "|m|": Math.sqrt(m[0] ** 2 + m[1] ** 2),
          "h.x": h[0], "h.y": h[1],
          "update.x": update[0], "update.y": update[1],
        } };
      },
      reset() {
        x = initialX;
        m = zero2();
        h = zero2();
      },
    };
  },
};
